### PR TITLE
feat(operator): Add new dragonfly k8s operator

### PR DESCRIPTION
**Pull request by Pothulapati**
This commit adds a new Kubernetes operator written in Golang into the repo, that can create and manage DragonFly Instances for Users. Operator is a tool to manage the lifecycle of DragonFly Instances in the cluster. It allows for easy creation, upgrade of dragonfly instances without having to understand the underlying details.

This commit of the operator only focuses on the installation part. More configurations, fail-over part will be added later.

## Running the Operator locally

The Operator can be ran locally by running the following commands.

```bash
kind create cluster

cd operator

make install

make run

kubectl apply -f config/samples/v1alpha1_dragonfly.yaml
```

You should see the Operator create the relevant
statefulset and service for the Dragonfly Instance.

```bash
kubectl get statefulsets
```

A lot of the boilerplate code here is generated by [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder)

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -83,6 +83,20 @@ jobs:
           ./multi_test --multi_exec_mode=2 --gtest_repeat=10
           ./multi_test --multi_exec_mode=3 --gtest_repeat=10
           # GLOG_logtostderr=1 GLOG_vmodule=transaction=1,engine_shard_set=1 CTEST_OUTPUT_ON_FAILURE=1 ninja server/test
+  build-operator:
```
Comment (Pothulapati) [inline .github/workflows/ci.yml:86]: separated this out as we don't need the matrix build that we have for dragonfly

File: operator/api/v1alpha1/dragonflydb_types.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	// Resources *corev1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+// DragonflyDbStatus defines the observed state of DragonflyDb
+type DragonflyDbStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Created is true if the database resources have been created
+	Created bool `json:"created,omitempty"`
```
Comment (Pothulapati) [inline operator/api/v1alpha1/dragonflydb_types.go:50]: `Created` will be replaced with `Phase` like we discussed (@ashotland), Ignore this for now. This is how the Operator if the resources were already created

File: operator/config/crd/bases/dragonflydb.io_dragonflydbs.yaml [hunk 3]
```diff
@@ -0,0 +1,57 @@
+---
+apiVersion: apiextensions.k8s.io/v1
```
Comment (Pothulapati) [inline operator/config/crd/bases/dragonflydb.io_dragonflydbs.yaml:2]: This is autogenerated based on the `dragonflydb_types.go` which is the main resources to change the schema of the `DragonFlyDb` resource.

File: operator/config/samples/v1alpha1_dragonflydb.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+apiVersion: dragonflydb.io/v1alpha1
```
Comment (Pothulapati) [inline operator/config/samples/v1alpha1_dragonflydb.yaml:1]: This is a sample of the schema that the user will create

File: operator/README.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
+# operator
```
[the remark this answers is not in the document]
Comment (ashotland) [reply operator/README.md:1]: If we are going with a separate repo  we can of course have it private to start with

**Review by ashotland — COMMENTED** [review]
Lots of "db" and "database" terminology

File: operator/config/manager/manager.yaml [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    app.kubernetes.io/name: namespace
+    app.kubernetes.io/instance: system
+    app.kubernetes.io/component: manager
+    app.kubernetes.io/created-by: operator
+    app.kubernetes.io/part-of: operator
+    app.kubernetes.io/managed-by: kustomize
+  name: system
+---
+apiVersion: apps/v1
+kind: Deployment
```
[the remark this answers is not in the document]
Comment (Pothulapati) [reply operator/config/manager/manager.yaml:15]: Correct!

**Comment by Pothulapati** [comment]
Closing this, as we now have a new repo! 👀

**Comment by DavidSpek** [comment]
@Pothulapati Is the new repo public? I’d love to have a look, test it and possibly contribute to it as needed.

**Comment by DavidSpek** [comment]
I was personally already thinking an operator would be a good fit to handle failover for HA deployments, similar to what the zalando Postgres operator does. So I guess that would be my main use-case. In general I was looking at dragonflydb as a possibly interesting app to onboard onto our (open-source) marketplace https://www.plural.sh/. We have a number of apps that require Redis and this week I’ll be working on one that needs Memcached (Grafana Mimir), but dragonflydb seems like a potentially interesting alternative for those scenarios. I’m not if having dragonflydb on our marketplace would be allowed under the license though, but I’ll get in touch with you about that separately.

**Comment by Pothulapati** [comment]
Thanks, interesting!  - the operator will definitely cover master-replica HA functionality. let me know if you want me to connect you with the right folks to discuss licensing or commercial details regarding your marketplace.

I will ping you when we release the operator repo irrespective. :)

**Pull request closed without merging**